# Flask API

## How Flask works

When we open a browser we see a few things:
Front-End
- **HTML** displays the page elements like the actual text on the website
- **CSS** styles the elemtns liek change font or sizeof the text
- **Bootstrap** provides some automatic styling through CSS and Javascript

Every website will perform some main ops:
- Accept info from user
- Retrieve info from databse
- Create/Update/Delee info in the database
- Display information back to the user

To connect the front-end stuff to web database, we need a web framework.
WEb framework is to accept user information from website and connect to back-end database and report something releveant that cna be stored in db to the front-end to user can see it

**Flask** is a web framework. Allows to connect Python code to the web. Flask has reputation for being not scalable. 

## Accepting user information
- Could be something like accepting query into Google search - instance of user filling out info. WTForms is a library that works well with flask
- Communicate with database and retrieve information. In this case its SQLite. SQLite can scale to quite a large website.
- SQLAlchemy allows us to write Python code instead of SQL queries
- Flask is like connector between front-end and back-end info that is stored
- **Jinja** templates grab info from Python and Flask ad send info back as HTML
- Flask renders HTML templates

## Virtual Environment

Launched web app and want to use external python library, but the library gets updated. We can set up virtual environment to help manage dependencies. 

## Flask "Hello World" example
- Return a string "Hello Puppy" on a webpage

In [1]:
from flask import Flask 

#Next line creates an application object as an instance of the class flask
#name variable is a python pre-defined variable which is set to name of module in which it is used
#flask uses location of module as a starting point when loading other files
app = Flask(__name__) 

#linkes the page below to whatever route on web ap it should be at
#'/' is same thing as single home page or domain
@app.route('/')

#defines a page index with the function and returns a string
def index():
    return '<h1>Hello Puppy!</h1>'

#if running script, run the application
if __name__ == '__main__':
    app.run()

 * Serving Flask app "__main__" (lazy loading)
 * Environment: production
   Use a production WSGI server instead.
 * Debug mode: off


 * Running on http://127.0.0.1:5000/ (Press CTRL+C to quit)
127.0.0.1 - - [15/Dec/2019 15:09:01] "GET / HTTP/1.1" 200 -
127.0.0.1 - - [15/Dec/2019 15:09:23] "GET /some HTTP/1.1" 404 -
127.0.0.1 - - [15/Dec/2019 15:09:33] "GET / HTTP/1.1" 200 -


## Basic Routing

- @app.route() is the main functon
- Home domain is represented as http://127.0.0.1:5000/
- @app.route("/some_page") is a new page
- Once deploted to web, its going to be replaced by the domain name

In [2]:
from flask import Flask 

app = Flask(__name__) 

@app.route('/')
def index():
    return '<h1>Hello Puppy!</h1>'

@app.route('/info') #this is just defining anothe rpage
def info():
    return "<h1> Puppies are cute!</h1>"

if __name__ == '__main__':
    app.run()

 * Serving Flask app "__main__" (lazy loading)
 * Environment: production
   Use a production WSGI server instead.
 * Debug mode: off


 * Running on http://127.0.0.1:5000/ (Press CTRL+C to quit)
127.0.0.1 - - [15/Dec/2019 15:13:46] "GET / HTTP/1.1" 200 -
127.0.0.1 - - [15/Dec/2019 15:13:52] "GET /info HTTP/1.1" 200 -


## Server defintion

Method | Description
-- | --
GET | The browser tells the server to just get the information stored on that page and send it. This is probably the most common method.
HEAD | The browser tells the server to get the information, but it is only interested in the headers, not the content of the page. An application is supposed to handle that as if a GET request was received but to not deliver the actual content. In Flask you don’t have to deal with that at all, the underlying Werkzeug library handles that for you.
POST | The browser tells the server that it wants to post some new information to that URL and that the server must ensure the data is stored and only stored once. This is how HTML forms usually transmit data to the server.
PUT | Similar to POST but the server might trigger the store procedure multiple times by overwriting the old values more than once. Now you might be asking why this is useful, but there are some good reasons to do it this way. Consider that the connection is lost during transmission: in this situation a system between the browser and the server might receive the request safely a second time without breaking things. With POST that would not be possible because it must only be triggered once.
DELETE | Remove the information at the given location.
OPTIONS | Provides a quick way for a client to figure out which methods are supported by this URL. Starting with Flask 0.6, this is implemented for you automatically.

**Response Code**

HTTP Code | Response Data
--|--
200 | The request has succeeded. 
400 | Bad Request.
404 | Not Found.

In [3]:
from flask import Flask, jsonify,request
import joblib
import pandas as pd

In [ ]:
app = Flask(__name__)

@app.route('/',methods=['POST'])
def home():
    return 'Hello World'

@app.route('/predict', methods=['POST'])
def predict():
    json_ = request.get_json(silent=True)
    message=json_.get('message')
    mydf = pd.DataFrame({'message':message})
    print(mydf)
    prediction = clf.predict_proba(mydf['message'])
    return jsonify({'prediction': prediction.tolist()})

## Running the server

In [ ]:
# mymodel=open('my_model_pipeline.pkl','rb')
# clf=joblib.load(mymodel)
# app.run(port=5000)

## Prediction from server

In [4]:
import os
import json
import requests

In [5]:
payload = {'message' : ["FREE FREE FREE"]}
r = requests.post("http://127.0.0.1:5000/predict" , json = payload)
r

<Response [200]>

In [6]:
r.text

'{"prediction":[[0.00849832658995884,0.9915016734100409]]}\n'

In [7]:
res = r.json()

In [8]:
res["prediction"]

[[0.00849832658995884, 0.9915016734100409]]